internal-packages/run-engine/src/engine/systems/executionSnapshotSystem.ts

```diff
@@ -141,6 +154,72 @@ export function executionResultFromSnapshot(snapshot: TaskRunExecutionSnapshot):
   };
 }
 
+export function executionDataFromSnapshot(snapshot: EnhancedExecutionSnapshot): RunExecutionData {
+  return {
+    version: "1" as const,
+    snapshot: {
+      id: snapshot.id,
+      friendlyId: snapshot.friendlyId,
+      executionStatus: snapshot.executionStatus,
+      description: snapshot.description,
+    },
+    run: {
+      id: snapshot.runId,
+      friendlyId: snapshot.runFriendlyId,
+      status: snapshot.runStatus,
+      attemptNumber: snapshot.attemptNumber ?? undefined,
+    },
+    batch: snapshot.batchId
+      ? {
+          id: snapshot.batchId,
+          friendlyId: BatchId.toFriendlyId(snapshot.batchId),
+        }
+      : undefined,
+    checkpoint: snapshot.checkpoint
+      ? {
+          id: snapshot.checkpoint.id,
+          friendlyId: snapshot.checkpoint.friendlyId,
+          type: snapshot.checkpoint.type,
+          location: snapshot.checkpoint.location,
+          imageRef: snapshot.checkpoint.imageRef,
+          reason: snapshot.checkpoint.reason ?? undefined,
+        }
+      : undefined,
+    completedWaitpoints: snapshot.completedWaitpoints,
+  };
+}
+
+export async function getExecutionSnapshotsSince(
+  prisma: PrismaClientOrTransaction,
+  runId: string,
+  sinceSnapshotId: string
+): Promise<EnhancedExecutionSnapshot[]> {
+  // Find the createdAt of the sinceSnapshotId
+  const sinceSnapshot = await prisma.taskRunExecutionSnapshot.findUnique({
+    where: { id: sinceSnapshotId },
+    select: { createdAt: true },
+  });
+
+  if (!sinceSnapshot) {
+    throw new Error(`No execution snapshot found for id ${sinceSnapshotId}`);
+  }
+
+  const snapshots = await prisma.taskRunExecutionSnapshot.findMany({
+    where: {
+      runId,
+      isValid: true,
+      createdAt: { gt: sinceSnapshot.createdAt },
+    },
+    include: {
+      completedWaitpoints: true,
+      checkpoint: true,
+    },
+    orderBy: { createdAt: "asc" },
+  });
+
+  return snapshots.map(enhanceExecutionSnapshot);
+}
+
```

Comment:
This makes me think we should have an index with the id instead of createdAt (or as well as)